File: prob2.java
// Time Complexity : O(n)
// Space Complexity : O(n * m)
// Did this code successfully run on Leetcode : Yes


class Solution {
    public int change(int amount, int[] coins) {

        int[][] dp = new int[coins.length+1][amount+1];
        dp[0][0]=1;
        for (int i=1; i<=coins.length; i++)
        {
             dp[i][0] = 1;
            for(int j=1; j<=amount; j++)
            {
                dp[i][j] = dp[i - 1][j];
                if(coins[i-1]<=j)
                {
                    dp[i][j]+=dp[i][j-coins[i-1]];
                }
            }
        }
        return dp[coins.length][amount];
    }
}

File: prob2.py
# Time Complexity : O(n*amount)
# Space Complexity :O(n*amount)
# Passed on Leetcode: yes

class Solution:
    def change(self, amount: int, coins: List[int]) -> int:

        nCoins = len(coins)
        dp = [[None for _ in range(amount+1)] for _ in range(nCoins+1)]
        
        for i in range(1, amount+1):
            dp[0][i] = 0
        for i in range(nCoins+1):
            dp[i][0] = 1
        
        for i in range(1, nCoins+1):
            for j in range(1, amount+1):
                if coins[i-1] > j:
                    dp[i][j] = dp[i-1][j]
                else:
                    dp[i][j] = dp[i-1][j] + dp[i][j-coins[i-1]]

        return dp[nCoins][amount]
        